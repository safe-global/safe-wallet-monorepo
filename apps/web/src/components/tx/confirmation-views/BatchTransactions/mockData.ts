import type { DraftBatchItem } from '@/store/batchSlice'
import { OperationType } from '@safe-global/types-kit'

export const mockedDraftBatch: DraftBatchItem[] = [
  {
    id: '6283sw7pzyk',
    timestamp: 1726820415651,
    txData: {
      to: '0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6',
      value: '1000000000000',
      data: '0x',
      operation: OperationType.Call,
    },
  },
  {
    id: 'abc123def456',
    timestamp: 1726820415652,
    txData: {
      to: '0x1234567890123456789012345678901234567890',
      value: '500000000000',
      data: '0x',
      operation: OperationType.Call,
    },
  },
]
