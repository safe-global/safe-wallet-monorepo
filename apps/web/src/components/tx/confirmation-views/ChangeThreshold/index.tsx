import { Box, Divider, Typography } from '@mui/material'
import React from 'react'

import commonCss from '@/components/tx-flow/common/styles.module.css'
import useSafeInfo from '@/hooks/useSafeInfo'
import { ChangeSignerSetupWarning } from '@/features/multichain/components/SignerSetupWarning/ChangeSignerSetupWarning'
import { type TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
import { isChangeThresholdView } from '../utils'
import { maybePlural } from '@safe-global/utils/utils/formatters'

interface ChangeThresholdProps {
  txInfo?: TransactionDetails['txInfo']
}

function ChangeThreshold({ txInfo }: ChangeThresholdProps) {
  const { safe } = useSafeInfo()
  const threshold = txInfo && isChangeThresholdView(txInfo) && txInfo.settingsInfo?.threshold

  return (
    <>
      <ChangeSignerSetupWarning />

      <div>
        <Typography variant="body2" color="text.secondary" mb={0.5}>
          Any transaction will require the confirmation of:
        </Typography>

        <Typography aria-label="threshold">
          <b>{threshold}</b> out of{' '}
          <b>
            {safe.owners.length} signer{maybePlural(safe.owners)}
          </b>
        </Typography>
      </div>
      <Box
        sx={{
          my: 1,
        }}
      >
        <Divider className={commonCss.nestedDivider} />
      </Box>
    </>
  )
}

export default ChangeThreshold
