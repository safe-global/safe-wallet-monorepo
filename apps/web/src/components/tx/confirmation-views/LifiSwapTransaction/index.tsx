import { DataTable } from '@/components/common/Table/DataTable'
import { Stack, Typography } from '@mui/material'
import { type SwapTransactionInfo } from '@safe-global/store/gateway/AUTO_GENERATED/transactions'
import { formatUnits } from 'ethers'
import SwapTokens from '@/features/swap/components/SwapTokens'
import NamedAddressInfo from '@/components/common/NamedAddressInfo'
import { DataRow } from '@/components/common/Table/DataRow'
import { formatAmount } from '@safe-global/utils/utils/formatNumber'
import TokenAmount from '@/components/common/TokenAmount'
import ExternalLink from '@/components/common/ExternalLink'
import css from './styles.module.css'

const PreviewSwapAmount = ({ txInfo }: { txInfo: SwapTransactionInfo }) => (
  <div key="amount">
    <SwapTokens
      first={{
        value: txInfo.fromAmount,
        label: 'Sell',
        tokenInfo: txInfo.fromToken,
      }}
      second={{
        value: txInfo.toAmount,
        label: 'For at least',
        tokenInfo: txInfo.toToken,
      }}
    />
  </div>
)

const ListSwapAmount = ({ txInfo }: { txInfo: SwapTransactionInfo }) => (
  <DataRow datatestid="amount" key="amount" title="Amount">
    <Stack spacing={0.5}>
      <Typography display="flex" alignItems="center" flexDirection="row" gap={1}>
        Sell{' '}
        <TokenAmount
          value={txInfo.fromAmount}
          decimals={txInfo.fromToken.decimals}
          logoUri={txInfo.fromToken.logoUri ?? ''}
          tokenSymbol={txInfo.fromToken.symbol}
        />
      </Typography>
      <Typography display="flex" alignItems="center" flexDirection="row" gap={1}>
        For{' '}
        <TokenAmount
          value={txInfo.toAmount}
          decimals={txInfo.toToken.decimals}
          logoUri={txInfo.toToken.logoUri ?? ''}
          tokenSymbol={txInfo.toToken.symbol}
        />
      </Typography>
    </Stack>
  </DataRow>
)

export const LifiSwapTransaction = ({ txInfo, isPreview }: { txInfo: SwapTransactionInfo; isPreview: boolean }) => {
  const totalFee = formatUnits(
    BigInt(txInfo.fees?.integratorFee ?? 0n) + BigInt(txInfo.fees?.lifiFee ?? 0n),
    txInfo.fromToken.decimals,
  )

  const fromAmountDecimals = formatUnits(txInfo.fromAmount, txInfo.fromToken.decimals)
  const toAmountDecimals = formatUnits(txInfo.toAmount, txInfo.toToken.decimals)
  const exchangeRate = Number(toAmountDecimals) / Number(fromAmountDecimals)

  const rows = [
    isPreview ? <PreviewSwapAmount txInfo={txInfo} /> : <ListSwapAmount txInfo={txInfo} />,
    <DataRow datatestid="price" key="price" title="Price">
      1 {txInfo.fromToken.symbol} = {formatAmount(exchangeRate)} {txInfo.toToken!.symbol}
    </DataRow>,
    <DataRow datatestid="receiver" key="Receiver" title="Receiver">
      <NamedAddressInfo
        address={txInfo.recipient.value}
        name={txInfo.recipient.name}
        hasExplorer
        showAvatar={false}
        onlyName
        showCopyButton
      />
    </DataRow>,
    <DataRow datatestid="total-fee" key="fees" title="Fees">
      {formatAmount(totalFee)} {txInfo.fromToken.symbol}
    </DataRow>,
  ]

  if (txInfo.lifiExplorerUrl) {
    rows.push(
      <DataRow datatestid="lifi-explorer-url" key="lifi-explorer-url" title="Lifi Explorer">
        <ExternalLink className={css.externalLink} href={txInfo.lifiExplorerUrl}>
          View in LiFi explorer
        </ExternalLink>
      </DataRow>,
    )
  }

  return (
    <Stack>
      <DataTable rows={rows} />
    </Stack>
  )
}
