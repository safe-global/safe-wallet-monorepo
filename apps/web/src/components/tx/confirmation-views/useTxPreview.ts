import { Operation, getTxPreview } from '@safe-global/safe-gateway-typescript-sdk'
import type { SafeTransaction } from '@safe-global/safe-core-sdk-types'
import useAsync from '@/hooks/useAsync'
import useSafeInfo from '@/hooks/useSafeInfo'
import { sameAddress } from '@/utils/addresses'
import { Interface } from 'ethers'

export const InvalidPreviewErrorName = 'InvalidPreviewError'
class InvalidPreviewError extends Error {
  constructor(message: string) {
    super(message)

    this.name = InvalidPreviewErrorName
  }
}

const useTxPreview = (
  safeTxData?: {
    operation: SafeTransaction['data']['operation']
    data: SafeTransaction['data']['data']
    value: SafeTransaction['data']['value']
    to: SafeTransaction['data']['to']
  },
  customSafeAddress?: string,
  txId?: string,
) => {
  const skip = !!txId || !safeTxData
  const {
    safe: { chainId },
    safeAddress,
  } = useSafeInfo()
  const address = customSafeAddress ?? safeAddress
  const { operation = Operation.CALL, data = '', to, value } = safeTxData ?? {}

  return useAsync(async () => {
    if (skip) return
    const txPreview = await getTxPreview(chainId, address, operation, data, to, value)

    // Validate that the tx preview matches the txData
    // First of all the top level data needs to match:
    const matchesSafeTxData =
      sameAddress(txPreview.txData.to.value, safeTxData.to) &&
      Number(txPreview.txData.operation) === Number(safeTxData.operation) &&
      txPreview.txData.value === safeTxData.value &&
      (txPreview.txData.hexData ?? '0x') === safeTxData.data

    if (!matchesSafeTxData) {
      throw new InvalidPreviewError("SafeTx data does not match the preview result's transaction data")
    }

    // validate the decodedData
    const dataDecoded = txPreview.txData.dataDecoded

    if (dataDecoded) {
      const abiString = `function ${dataDecoded.method}(${dataDecoded.parameters?.map((param) => param.type).join(',')})`
      const abiInterface = new Interface([abiString])
      const rawDataFromDecodedData = abiInterface.encodeFunctionData(
        dataDecoded.method,
        dataDecoded.parameters?.map((param) => param.value),
      )

      if (rawDataFromDecodedData !== safeTxData.data) {
        throw new InvalidPreviewError('Decoded data does not match raw data')
      }
    }

    return txPreview
  }, [skip, chainId, address, operation, data, to, value, safeTxData])
}

export default useTxPreview
