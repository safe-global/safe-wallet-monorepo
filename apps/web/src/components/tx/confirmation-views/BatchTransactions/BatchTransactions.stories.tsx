import type { Meta, StoryObj } from '@storybook/react'
import { Paper, ThemeProvider } from '@mui/material'
import { StoreDecorator } from '@/stories/storeDecorator'
import BatchTransactions from './index'
import { mockedDraftBatch } from './mockData'
import createSafeTheme from '@/components/theme/safeTheme'

const meta = {
  component: BatchTransactions,
  parameters: {
    layout: 'centered',
  },
  decorators: [
    (Story) => {
      return (
        <StoreDecorator
          initialState={{
            chains: { data: [{ chainId: '11155111' }] },
            batch: {
              '11155111': {
                '': [mockedDraftBatch[0], mockedDraftBatch[0]],
              },
            },
          }}
        >
          <ThemeProvider theme={createSafeTheme('dark')}>
            <Paper sx={{ padding: 2 }}>
              <Story />
            </Paper>
          </ThemeProvider>
        </StoreDecorator>
      )
    },
  ],

  tags: ['autodocs'],
} satisfies Meta<typeof BatchTransactions>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {}
