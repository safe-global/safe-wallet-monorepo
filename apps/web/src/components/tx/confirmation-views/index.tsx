import type { TransactionPreview } from '@safe-global/safe-gateway-typescript-sdk'
import { type TransactionDetails } from '@safe-global/safe-gateway-typescript-sdk'
import type { SafeTransaction } from '@safe-global/safe-core-sdk-types'
import {
  isAnyStakingTxInfo,
  isCustomTxInfo,
  isExecTxData,
  isOnChainConfirmationTxData,
  isOnChainSignMessageTxData,
  isSafeMigrationTxData,
  isSafeUpdateTxData,
  isSwapOrderTxInfo,
  isTwapOrderTxInfo,
  isVaultDepositTxInfo,
  isVaultRedeemTxInfo,
} from '@/utils/transaction-guards'
import { type ReactNode, useContext, useMemo, useRef, useState, useEffect } from 'react'
import type { NarrowConfirmationViewProps } from './types'
import SettingsChange from './SettingsChange'
import ChangeThreshold from './ChangeThreshold'
import BatchTransactions from './BatchTransactions'
import { TxModalContext } from '@/components/tx-flow'
import { isSettingsChangeView, isChangeThresholdView, isConfirmBatchView } from './utils'
import { OnChainConfirmation } from '@/components/transactions/TxDetails/TxData/NestedTransaction/OnChainConfirmation'
import { ExecTransaction } from '@/components/transactions/TxDetails/TxData/NestedTransaction/ExecTransaction'
import { type ReactElement } from 'react'
import SwapOrder from './SwapOrder'
import StakingTx from './StakingTx'
import UpdateSafe from './UpdateSafe'
import { MigrateToL2Information } from './MigrateToL2Information'
import { NestedSafeCreation } from './NestedSafeCreation'
import { isNestedSafeCreation } from '@/utils/nested-safes'
import VaultDepositConfirmation from 'src/features/earn/components/VaultDepositConfirmation'
import VaultRedeemConfirmation from '@/features/earn/components/VaultRedeemConfirmation'
import Summary from '@/components/transactions/TxDetails/Summary'
import TxData from '@/components/transactions/TxDetails/TxData'
import { isMultiSendCalldata } from '@/utils/transaction-calldata'
import useChainId from '@/hooks/useChainId'
import { TransactionWarnings } from '../TransactionWarnings'
import { Box } from '@mui/material'
import DecodedData from '@/components/transactions/TxDetails/TxData/DecodedData'

type ConfirmationViewProps = {
  txDetails?: TransactionDetails
  txPreview?: TransactionPreview
  safeTx?: SafeTransaction
  isBatch?: boolean
  isApproval?: boolean
  isCreation?: boolean
  children?: ReactNode
}

const getConfirmationViewComponent = ({
  txInfo,
  txData,
  txFlow,
}: NarrowConfirmationViewProps & { txFlow?: ReactElement }) => {
  if (isChangeThresholdView(txInfo)) return <ChangeThreshold txInfo={txInfo} />

  if (isConfirmBatchView(txFlow)) return <BatchTransactions />

  if (isSettingsChangeView(txInfo)) return <SettingsChange txInfo={txInfo} />

  if (isOnChainConfirmationTxData(txData)) return <OnChainConfirmation data={txData} isConfirmationView />

  if (isExecTxData(txData)) return <ExecTransaction data={txData} isConfirmationView />

  if (isSwapOrderTxInfo(txInfo) || isTwapOrderTxInfo(txInfo)) return <SwapOrder txInfo={txInfo} txData={txData} />

  if (isAnyStakingTxInfo(txInfo)) return <StakingTx txInfo={txInfo} />

  // @ts-expect-error TODO: Fix these if there is time
  if (isVaultDepositTxInfo(txInfo)) return <VaultDepositConfirmation txInfo={txInfo} />

  // @ts-expect-error TODO: Fix these if there is time
  if (isVaultRedeemTxInfo(txInfo)) return <VaultRedeemConfirmation txInfo={txInfo} />

  if (isCustomTxInfo(txInfo) && isSafeUpdateTxData(txData)) return <UpdateSafe txData={txData} />

  if (isCustomTxInfo(txInfo) && isSafeMigrationTxData(txData)) {
    return <MigrateToL2Information variant="queue" txData={txData} />
  }

  if (isCustomTxInfo(txInfo) && txData && isNestedSafeCreation(txData)) {
    return <NestedSafeCreation txData={txData} />
  }

  return null
}

const ConfirmationView = ({ safeTx, txPreview, txDetails, ...props }: ConfirmationViewProps) => {
  const { txFlow } = useContext(TxModalContext)
  const details = txDetails ?? txPreview
  const chainId = useChainId()

  // Used to check if the decoded data was rendered inside the TxData component
  // If it was, we hide the decoded data in the Summary to avoid showing it twice
  const decodedDataRef = useRef(null)
  const [isDecodedDataVisible, setIsDecodedDataVisible] = useState(false)

  useEffect(() => {
    // If decodedDataRef.current is not null, the decoded data was rendered inside the TxData component
    setIsDecodedDataVisible(!!decodedDataRef.current)
  }, [])

  const ConfirmationViewComponent = useMemo(() => {
    return details
      ? getConfirmationViewComponent({
          txInfo: details.txInfo,
          txData: details.txData,
          txFlow,
        })
      : undefined
  }, [details, txFlow])

  const showTxDetails =
    details !== undefined &&
    !isMultiSendCalldata(details.txData?.hexData ?? '0x') &&
    !isOnChainSignMessageTxData(details?.txData, chainId)

  return (
    <>
      <TransactionWarnings txData={details?.txData} />
      {ConfirmationViewComponent ||
        (details && showTxDetails && (
          <TxData txData={details?.txData} txInfo={details?.txInfo} txDetails={txDetails} imitation={false} trusted>
            <Box ref={decodedDataRef}>
              <DecodedData
                txData={details.txData}
                toInfo={isCustomTxInfo(details.txInfo) ? details.txInfo.to : details.txData?.to}
              />
            </Box>
          </TxData>
        ))}

      {props.children}

      <Summary
        safeTxData={safeTx?.data}
        txDetails={txDetails}
        txData={details?.txData}
        txInfo={details?.txInfo}
        showDecodedData={!isDecodedDataVisible}
      />
    </>
  )
}

export default ConfirmationView
