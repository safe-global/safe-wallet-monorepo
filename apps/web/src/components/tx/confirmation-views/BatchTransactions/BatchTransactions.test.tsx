import { render } from '@/tests/test-utils'
import BatchTransactions from '.'
import * as useDraftBatch from '@/hooks/useDraftBatch'
import * as useTxPreview from '@/components/tx/confirmation-views/useTxPreview'
import { mockedDraftBatch } from './mockData'
import {
  type TransactionPreview,
  Operation,
  TransactionInfoType,
  TransferDirection,
  TransactionTokenType,
} from '@safe-global/safe-gateway-typescript-sdk'

jest.spyOn(useDraftBatch, 'useDraftBatch').mockImplementation(() => mockedDraftBatch)

const mockUseTxPreview = jest.spyOn(useTxPreview, 'default')

const mockTxPreview: TransactionPreview = {
  txData: {
    hexData: '0x',
    dataDecoded: undefined,
    to: { value: '0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6', name: 'GnosisSafeProxy', logoUri: undefined },
    value: '1000000000000',
    operation: Operation.CALL,
    trustedDelegateCallTarget: false,
    addressInfoIndex: {
      '0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6': {
        value: '0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6',
        name: 'GnosisSafeProxy',
        logoUri: undefined,
      },
    },
  },
  txInfo: {
    type: TransactionInfoType.TRANSFER,
    humanDescription: undefined,
    sender: { value: '0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6', name: undefined, logoUri: undefined },
    recipient: { value: '0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6', name: 'GnosisSafeProxy', logoUri: undefined },
    direction: TransferDirection.OUTGOING,
    transferInfo: { type: TransactionTokenType.NATIVE_COIN, value: '1000000000000' },
  },
}

describe('BatchTransactions', () => {
  beforeEach(() => {
    jest.clearAllMocks()

    mockUseTxPreview.mockImplementation(() => [mockTxPreview, undefined, false] as any)
  })

  it('should render a list of batch transactions', () => {
    const { container, getByText } = render(<BatchTransactions />)

    expect(container).toMatchSnapshot()
    expect(getByText('GnosisSafeProxy')).toBeDefined()
  })
})
