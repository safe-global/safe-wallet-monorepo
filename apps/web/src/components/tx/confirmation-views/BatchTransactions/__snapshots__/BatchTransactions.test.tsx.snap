// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchTransactions should render a list of batch transactions 1`] = `
<div>
  <ul
    class="MuiList-root MuiList-padding css-8g50lk-MuiList-root"
  >
    <li
      class="MuiListItem-root MuiListItem-gutters css-rqnn2j-MuiListItem-root"
    >
      <div
        class="number"
      >
        1
      </div>
      <div
        class="accordion"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1gdgqta-MuiPaper-root-MuiAccordion-root"
          style="--Paper-shadow: none;"
        >
          <h3
            class="MuiAccordion-heading css-cy7rkm-MuiAccordion-heading"
          >
            <button
              aria-expanded="false"
              class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters accordion css-10sjung-MuiButtonBase-root-MuiAccordionSummary-root"
              data-testid="action-item"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-1r0e0ir-MuiAccordionSummary-content"
              >
                <div
                  class="summary"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-gvpe62-MuiSvgIcon-root"
                    data-testid="CodeIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M9.4 16.6 4.8 12l4.6-4.6L8 6l-6 6 6 6zm5.2 0 4.6-4.6-4.6-4.6L16 6l6 6-6 6z"
                    />
                  </svg>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-v6lhhw-MuiTypography-root"
                  />
                  <div
                    class="MuiStack-root css-kdc3n8-MuiStack-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-v6lhhw-MuiTypography-root"
                    >
                      Send
                    </p>
                    <span
                      aria-label="0.000001 ETH"
                      class="container"
                      data-mui-internal-clone-element="true"
                    >
                      <b
                        class="tokenText"
                      >
                        &lt; 0.00001
                         
                        ETH
                      </b>
                    </span>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-v6lhhw-MuiTypography-root"
                    >
                      to
                    </p>
                    <div
                      class="container"
                    >
                      <div
                        class="avatarContainer"
                        style="width: 16px; height: 16px;"
                      >
                        <div
                          class="icon"
                          style="background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA4IDgiIHNoYXBlLXJlbmRlcmluZz0ib3B0aW1pemVTcGVlZCIgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0Ij48cGF0aCBmaWxsPSJoc2woMjI2IDc1JSA0NCUpIiBkPSJNMCwwSDhWOEgweiIvPjxwYXRoIGZpbGw9ImhzbCg3MCA1NyUgNTYlKSIgZD0iTTAsMGgxdjFoLTF6TTcsMGgxdjFoLTF6TTEsMGgxdjFoLTF6TTYsMGgxdjFoLTF6TTIsMGgxdjFoLTF6TTUsMGgxdjFoLTF6TTAsMWgxdjFoLTF6TTcsMWgxdjFoLTF6TTAsMmgxdjFoLTF6TTcsMmgxdjFoLTF6TTEsMmgxdjFoLTF6TTYsMmgxdjFoLTF6TTIsMmgxdjFoLTF6TTUsMmgxdjFoLTF6TTMsMmgxdjFoLTF6TTQsMmgxdjFoLTF6TTEsM2gxdjFoLTF6TTYsM2gxdjFoLTF6TTAsNGgxdjFoLTF6TTcsNGgxdjFoLTF6TTEsNWgxdjFoLTF6TTYsNWgxdjFoLTF6TTMsNWgxdjFoLTF6TTQsNWgxdjFoLTF6TTEsNmgxdjFoLTF6TTYsNmgxdjFoLTF6TTIsNmgxdjFoLTF6TTUsNmgxdjFoLTF6TTAsN2gxdjFoLTF6TTcsN2gxdjFoLTF6Ii8+PHBhdGggZmlsbD0iaHNsKDMzMSA4NiUgNDglKSIgZD0iTTAsM2gxdjFoLTF6TTcsM2gxdjFoLTF6TTIsNWgxdjFoLTF6TTUsNWgxdjFoLTF6TTMsNmgxdjFoLTF6TTQsNmgxdjFoLTF6TTIsN2gxdjFoLTF6TTUsN2gxdjFoLTF6Ii8+PC9zdmc+); width: 16px; height: 16px;"
                        />
                      </div>
                      <div
                        class="inline MuiBox-root css-1lchl8k"
                      >
                        <div
                          class="addressContainer inline"
                        >
                          <div
                            class="MuiBox-root css-b5p5gz"
                          >
                            <span>
                              0xA77D...98b6
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </span>
              <span
                class="MuiAccordionSummary-expandIconWrapper css-1wqf3nl-MuiAccordionSummary-expandIconWrapper"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1dhtbeh-MuiSvgIcon-root"
                  data-testid="ExpandMoreIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
            </button>
          </h3>
          <div
            class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-cwrbtg-MuiCollapse-root"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper MuiCollapse-vertical css-1x6hinx-MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner MuiCollapse-vertical css-1i4ywhz-MuiCollapse-wrapperInner"
              >
                <div
                  class="MuiAccordion-region"
                  role="region"
                >
                  <div
                    class="MuiAccordionDetails-root css-w74p4c-MuiAccordionDetails-root"
                  >
                    <div
                      class="MuiStack-root css-jfdv4h-MuiStack-root"
                    >
                      <div
                        class="MuiStack-root css-1sazv7p-MuiStack-root"
                      >
                        <div
                          class="MuiGrid-root MuiGrid-container css-1yff1ei-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-ezmk0c-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <span
                              class="MuiTypography-root MuiTypography-body2 css-1ew0eu5-MuiTypography-root"
                            >
                              Interacted with
                            </span>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            <div
                              class="MuiTypography-root MuiTypography-body2 css-17vdyq3-MuiTypography-root"
                            >
                              <div
                                class="container"
                              >
                                <div
                                  class="avatarContainer"
                                  style="width: 20px; height: 20px;"
                                >
                                  <div
                                    class="icon"
                                    style="background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA4IDgiIHNoYXBlLXJlbmRlcmluZz0ib3B0aW1pemVTcGVlZCIgd2lkdGg9IjY0IiBoZWlnaHQ9IjY0Ij48cGF0aCBmaWxsPSJoc2woMjI2IDc1JSA0NCUpIiBkPSJNMCwwSDhWOEgweiIvPjxwYXRoIGZpbGw9ImhzbCg3MCA1NyUgNTYlKSIgZD0iTTAsMGgxdjFoLTF6TTcsMGgxdjFoLTF6TTEsMGgxdjFoLTF6TTYsMGgxdjFoLTF6TTIsMGgxdjFoLTF6TTUsMGgxdjFoLTF6TTAsMWgxdjFoLTF6TTcsMWgxdjFoLTF6TTAsMmgxdjFoLTF6TTcsMmgxdjFoLTF6TTEsMmgxdjFoLTF6TTYsMmgxdjFoLTF6TTIsMmgxdjFoLTF6TTUsMmgxdjFoLTF6TTMsMmgxdjFoLTF6TTQsMmgxdjFoLTF6TTEsM2gxdjFoLTF6TTYsM2gxdjFoLTF6TTAsNGgxdjFoLTF6TTcsNGgxdjFoLTF6TTEsNWgxdjFoLTF6TTYsNWgxdjFoLTF6TTMsNWgxdjFoLTF6TTQsNWgxdjFoLTF6TTEsNmgxdjFoLTF6TTYsNmgxdjFoLTF6TTIsNmgxdjFoLTF6TTUsNmgxdjFoLTF6TTAsN2gxdjFoLTF6TTcsN2gxdjFoLTF6Ii8+PHBhdGggZmlsbD0iaHNsKDMzMSA4NiUgNDglKSIgZD0iTTAsM2gxdjFoLTF6TTcsM2gxdjFoLTF6TTIsNWgxdjFoLTF6TTUsNWgxdjFoLTF6TTMsNmgxdjFoLTF6TTQsNmgxdjFoLTF6TTIsN2gxdjFoLTF6TTUsN2gxdjFoLTF6Ii8+PC9zdmc+); width: 20px; height: 20px;"
                                  />
                                </div>
                                <div
                                  class="MuiBox-root css-1lchl8k"
                                >
                                  <div
                                    class="ethHashInfo-name MuiBox-root css-171onha"
                                    title="GnosisSafeProxy"
                                  >
                                    <div
                                      class="MuiBox-root css-1o4wo1x"
                                    >
                                      GnosisSafeProxy
                                    </div>
                                  </div>
                                  <div
                                    class="addressContainer"
                                  >
                                    <div
                                      class="MuiBox-root css-b5p5gz"
                                    >
                                      <span
                                        aria-label="Copy to clipboard"
                                        class=""
                                        data-mui-internal-clone-element="true"
                                        style="cursor: pointer;"
                                      >
                                        <span>
                                          0xA77DE01e157f9f57C7c4A326eeE9C4874D0598b6
                                        </span>
                                      </span>
                                    </div>
                                    <span
                                      aria-label="Copy to clipboard"
                                      class=""
                                      data-mui-internal-clone-element="true"
                                      style="cursor: pointer;"
                                    >
                                      <button
                                        aria-label="Copy to clipboard"
                                        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-xyrzjn-MuiButtonBase-root-MuiIconButton-root"
                                        tabindex="0"
                                        type="button"
                                      >
                                        <mock-icon
                                          aria-hidden=""
                                          class="MuiSvgIcon-root MuiSvgIcon-colorBorder MuiSvgIcon-fontSizeSmall css-gvpe62-MuiSvgIcon-root"
                                          data-testid="copy-btn-icon"
                                          focusable="false"
                                        />
                                      </button>
                                    </span>
                                    <div
                                      class="MuiBox-root css-yjghm1"
                                    />
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiGrid-root MuiGrid-container css-1yff1ei-MuiGrid-root"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-item css-ezmk0c-MuiGrid-root"
                            data-testid="tx-row-title"
                            style="word-break: break-word;"
                          >
                            <span
                              class="MuiTypography-root MuiTypography-body2 css-1ew0eu5-MuiTypography-root"
                            >
                              Value
                            </span>
                          </div>
                          <div
                            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                            data-testid="tx-data-row"
                          >
                            <div
                              class="MuiBox-root css-axw7ok"
                            >
                              <img
                                alt="ETH"
                                class="image"
                                height="26"
                                loading="lazy"
                                referrerpolicy="no-referrer"
                              />
                              <p
                                class="MuiTypography-root MuiTypography-body2 css-1ct9qkn-MuiTypography-root"
                              >
                                ETH
                              </p>
                              <p
                                class="MuiTypography-root MuiTypography-body2 css-17vdyq3-MuiTypography-root"
                                data-testid="token-amount"
                              >
                                0.000001
                              </p>
                            </div>
                          </div>
                        </div>
                        <div
                          class="MuiTypography-root MuiTypography-body2 css-17vdyq3-MuiTypography-root"
                          data-testid="hexData"
                        >
                          <div
                            class="MuiGrid-root MuiGrid-container css-1yff1ei-MuiGrid-root"
                          >
                            <div
                              class="MuiGrid-root MuiGrid-item css-ezmk0c-MuiGrid-root"
                              data-testid="tx-row-title"
                              style="word-break: break-word;"
                            >
                              <span
                                class="MuiTypography-root MuiTypography-body2 css-1ew0eu5-MuiTypography-root"
                              >
                                Data
                              </span>
                            </div>
                            <div
                              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-1vd824g-MuiGrid-root"
                              data-testid="tx-data-row"
                            >
                              <div
                                class="encodedData MuiBox-root css-0"
                                data-testid="tx-hexData"
                              >
                                <span
                                  aria-label="Copy to clipboard"
                                  class=""
                                  data-mui-internal-clone-element="true"
                                  style="cursor: pointer;"
                                >
                                  <span
                                    class="monospace"
                                  >
                                    <b
                                      aria-label="The first 4 bytes determine the contract method that is being called"
                                      class=""
                                      data-mui-internal-clone-element="true"
                                    >
                                      0x
                                    </b>
                                     
                                  </span>
                                </span>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-gutters css-rqnn2j-MuiListItem-root"
    >
      <div
        class="number"
      >
        2
      </div>
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1daydue-MuiSkeleton-root"
        style="width: 100%; height: 56px;"
      />
    </li>
  </ul>
</div>
`;
